File: airbyte-cron/src/main/java/io/airbyte/cron/jobs/SelfHealTemporalWorkflows.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cron.jobs;

import static io.airbyte.cron.MicronautCronRunner.SCHEDULED_TRACE_OPERATION_NAME;

import datadog.trace.api.Trace;
import io.airbyte.commons.temporal.TemporalClient;
import io.airbyte.metrics.lib.MetricAttribute;
import io.airbyte.metrics.lib.MetricClient;
import io.airbyte.metrics.lib.MetricTags;
import io.airbyte.metrics.lib.OssMetricsRegistry;
import io.micronaut.scheduling.annotation.Scheduled;
import io.temporal.api.enums.v1.WorkflowExecutionStatus;
import jakarta.inject.Singleton;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Temporal cleaner. Resets failed workflow executions.
 */
@Singleton
public class SelfHealTemporalWorkflows {

  private static final Logger log = LoggerFactory.getLogger(SelfHealTemporalWorkflows.class);

  private final TemporalClient temporalClient;
  private final MetricClient metricClient;

  public SelfHealTemporalWorkflows(final TemporalClient temporalClient, final MetricClient metricClient) {
    log.debug("Creating temporal self-healing");
    this.temporalClient = temporalClient;
    this.metricClient = metricClient;
  }

  @Trace(operationName = SCHEDULED_TRACE_OPERATION_NAME)
  @Scheduled(fixedRate = "10s")
  void cleanTemporal() {
    metricClient.count(OssMetricsRegistry.CRON_JOB_RUN_BY_CRON_TYPE, 1, new MetricAttribute(MetricTags.CRON_TYPE, "self_heal_temporal"));
    final var numRestarted = temporalClient.restartClosedWorkflowByStatus(WorkflowExecutionStatus.WORKFLOW_EXECUTION_STATUS_FAILED);
    metricClient.count(OssMetricsRegistry.WORKFLOWS_HEALED, numRestarted);
  }

}

File: airbyte-cron/src/main/kotlin/io/airbyte/cron/jobs/SelfHealTemporalWorkflows.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cron.jobs

import datadog.trace.api.Trace
import io.airbyte.commons.temporal.TemporalClient
import io.airbyte.cron.SCHEDULED_TRACE_OPERATION_NAME
import io.airbyte.metrics.MetricAttribute
import io.airbyte.metrics.MetricClient
import io.airbyte.metrics.OssMetricsRegistry
import io.airbyte.metrics.lib.MetricTags
import io.github.oshai.kotlinlogging.KotlinLogging
import io.micronaut.scheduling.annotation.Scheduled
import io.temporal.api.enums.v1.WorkflowExecutionStatus
import jakarta.inject.Singleton

private val log = KotlinLogging.logger {}

@Singleton
class SelfHealTemporalWorkflows(
  private val temporalClient: TemporalClient,
  private val metricClient: MetricClient,
) {
  init {
    log.debug { "Creating temporal self-healing" }
  }

  @Trace(operationName = SCHEDULED_TRACE_OPERATION_NAME)
  @Scheduled(fixedRate = "10s")
  fun cleanTemporal() {
    metricClient.count(
      metric = OssMetricsRegistry.CRON_JOB_RUN_BY_CRON_TYPE,
      attributes = arrayOf(MetricAttribute(MetricTags.CRON_TYPE, "self_heal_temporal")),
    )

    temporalClient
      .restartClosedWorkflowByStatus(WorkflowExecutionStatus.WORKFLOW_EXECUTION_STATUS_FAILED)
      .also { metricClient.count(metric = OssMetricsRegistry.WORKFLOWS_HEALED, value = it.toLong()) }
  }
}
